File: code/cpp/ch19/tests/test_ch19.cpp
/*
 * Tests for Chapter 19: Graphs I — Exploring Networks
 * Build: g++ -std=c++17 -o /tmp/test_ch19 code/cpp/ch19/tests/test_ch19.cpp && /tmp/test_ch19
 */

#include <algorithm>
#include <cassert>
#include <iostream>
#include <queue>
#include <string>
#include <unordered_set>
#include <vector>
using namespace std;

// =====================================================================
// Reference solutions (uniquely named with ref_ prefix)
// =====================================================================

// --- W1: Build Adjacency List ---
vector<vector<int>> ref_build_adj_list(int n, vector<vector<int>> edges) {
    vector<vector<int>> adj(n);
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }
    for (auto& nbrs : adj) sort(nbrs.begin(), nbrs.end());
    return adj;
}

// --- W2: BFS Traversal ---
vector<int> ref_bfs_traversal(int n, vector<vector<int>> edges, int source) {
    vector<vector<int>> adj(n);
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }
    vector<bool> visited(n, false);
    visited[source] = true;
    queue<int> q;
    q.push(source);
    vector<int> order;
    while (!q.empty()) {
        int node = q.front(); q.pop();
        order.push_back(node);
        vector<int> nbrs = adj[node];
        sort(nbrs.begin(), nbrs.end());
        for (int nb : nbrs) {
            if (!visited[nb]) { visited[nb] = true; q.push(nb); }
        }
    }
    return order;
}

// --- W3: DFS Traversal ---
void ref_dfs_helper(vector<vector<int>>& adj, int node, vector<bool>& visited, vector<int>& order) {
    visited[node] = true;
    order.push_back(node);
    vector<int> nbrs = adj[node];
    sort(nbrs.begin(), nbrs.end());
    for (int nb : nbrs) {
        if (!visited[nb]) ref_dfs_helper(adj, nb, visited, order);
    }
}

vector<int> ref_dfs_traversal(int n, vector<vector<int>> edges, int source) {
    vector<vector<int>> adj(n);
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }
    vector<bool> visited(n, false);
    vector<int> order;
    ref_dfs_helper(adj, source, visited, order);
    return order;
}

// --- W4: Count Components ---
int ref_count_components(int n, vector<vector<int>> edges) {
    vector<vector<int>> adj(n);
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }
    vector<bool> visited(n, false);
    int count = 0;
    for (int v = 0; v < n; v++) {
        if (!visited[v]) {
            queue<int> q;
            q.push(v);
            visited[v] = true;
            while (!q.empty()) {
                int node = q.front(); q.pop();
                for (int nb : adj[node]) {
                    if (!visited[nb]) { visited[nb] = true; q.push(nb); }
                }
            }
            count++;
        }
    }
    return count;
}

// --- W5: Is Path Exists ---
bool ref_is_path_exists(int n, vector<vector<int>> edges, int source, int dest) {
    if (source == dest) return true;
    vector<vector<int>> adj(n);
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }
    vector<bool> visited(n, false);
    visited[source] = true;
    queue<int> q;
    q.push(source);
    while (!q.empty()) {
        int node = q.front(); q.pop();
        for (int nb : adj[node]) {
            if (nb == dest) return true;
            if (!visited[nb]) { visited[nb] = true; q.push(nb); }
        }
    }
    return false;
}

// --- P1: Shortest Path ---
vector<int> ref_shortest_path(int n, vector<vector<int>> edges, int source) {
    vector<vector<int>> adj(n);
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }
    vector<int> dist(n, -1);
    dist[source] = 0;
    queue<int> q;
    q.push(source);
    while (!q.empty()) {
        int node = q.front(); q.pop();
        for (int nb : adj[node]) {
            if (dist[nb] == -1) { dist[nb] = dist[node] + 1; q.push(nb); }
        }
    }
    return dist;
}

// --- P2: Detect Cycle ---
bool ref_dfs_cycle(vector<vector<int>>& adj, int node, int parent, vector<bool>& visited) {
    visited[node] = true;
    for (int nb : adj[node]) {
        if (!visited[nb]) {
            if (ref_dfs_cycle(adj, nb, node, visited)) return true;
        } else if (nb != parent) return true;
    }
    return false;
}

bool ref_detect_cycle(int n, vector<vector<int>> edges) {
    vector<vector<int>> adj(n);
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }
    vector<bool> visited(n, false);
    for (int v = 0; v < n; v++) {
        if (!visited[v] && ref_dfs_cycle(adj, v, -1, visited)) return true;
    }
    return false;
}

// --- P3: Bipartite Check ---
bool ref_bipartite(int n, vector<vector<int>> edges) {
    vector<vector<int>> adj(n);
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }
    vector<int> color(n, -1);
    for (int start = 0; start < n; start++) {
        if (color[start] != -1) continue;
        color[start] = 0;
        queue<int> q;
        q.push(start);
        while (!q.empty()) {
            int node = q.front(); q.pop();
            for (int nb : adj[node]) {
                if (color[nb] == -1) { color[nb] = 1 - color[node]; q.push(nb); }
                else if (color[nb] == color[node]) return false;
            }
        }
    }
    return true;
}

// --- P4: Clone Graph ---
vector<vector<int>> ref_clone_graph(vector<vector<int>> adj) {
    vector<vector<int>> clone;
    for (auto& nbrs : adj) clone.push_back(vector<int>(nbrs.begin(), nbrs.end()));
    return clone;
}

// --- P5: All Paths ---
void ref_all_paths_dfs(vector<vector<int>>& adj, int node, int target,
                        vector<int>& path, vector<vector<int>>& result) {
    if (node == target) { result.push_back(path); return; }
    vector<int> nbrs = adj[node];
    sort(nbrs.begin(), nbrs.end());
    for (int nb : nbrs) {
        path.push_back(nb);
        ref_all_paths_dfs(adj, nb, target, path, result);
        path.pop_back();
    }
}

vector<vector<int>> ref_all_paths(int n, vector<vector<int>> edges) {
    vector<vector<int>> adj(n);
    for (auto& e : edges) adj[e[0]].push_back(e[1]);
    vector<vector<int>> result;
    vector<int> path = {0};
    ref_all_paths_dfs(adj, 0, n - 1, path, result);
    sort(result.begin(), result.end());
    return result;
}

// --- C1: Number of Provinces ---
int ref_num_provinces(vector<vector<int>> isConnected) {
    int n = isConnected.size();
    vector<bool> visited(n, false);
    int count = 0;
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            queue<int> q;
            q.push(i);
            visited[i] = true;
            while (!q.empty()) {
                int city = q.front(); q.pop();
                for (int j = 0; j < n; j++) {
                    if (isConnected[city][j] == 1 && !visited[j]) {
                        visited[j] = true;
                        q.push(j);
                    }
                }
            }
            count++;
        }
    }
    return count;
}

// --- C2: Course Schedule ---
bool ref_has_cycle(vector<vector<int>>& adj, int node, vector<int>& state) {
    state[node] = 1;
    for (int nb : adj[node]) {
        if (state[nb] == 1) return true;
        if (state[nb] == 0 && ref_has_cycle(adj, nb, state)) return true;
    }
    state[node] = 2;
    return false;
}

bool ref_course_schedule(int numCourses, vector<vector<int>> prerequisites) {
    vector<vector<int>> adj(numCourses);
    for (auto& p : prerequisites) adj[p[1]].push_back(p[0]);
    vector<int> state(numCourses, 0);
    for (int c = 0; c < numCourses; c++) {
        if (state[c] == 0 && ref_has_cycle(adj, c, state)) return false;
    }
    return true;
}

// --- C3: Word Ladder ---
int ref_word_ladder(string beginWord, string endWord, vector<string> wordList) {
    unordered_set<string> wordSet(wordList.begin(), wordList.end());
    if (!wordSet.count(endWord)) return 0;
    queue<pair<string, int>> q;
    q.push({beginWord, 1});
    unordered_set<string> visited;
    visited.insert(beginWord);
    while (!q.empty()) {
        auto [word, length] = q.front(); q.pop();
        if (word == endWord) return length;
        for (int i = 0; i < (int)word.size(); i++) {
            char original = word[i];
            for (char c = 'a'; c <= 'z'; c++) {
                if (c == original) continue;
                word[i] = c;
                if (wordSet.count(word) && !visited.count(word)) {
                    visited.insert(word);
                    q.push({word, length + 1});
                }
            }
            word[i] = original;
        }
    }
    return 0;
}

// =====================================================================
// Test runner
// =====================================================================

int passed = 0, failed = 0;

void check(bool condition, const string& msg) {
    if (condition) { passed++; }
    else { failed++; cout << "FAIL: " << msg << endl; }
}

int main() {
    cout << "Chapter 19: Graphs I — Exploring Networks" << endl;
    cout << "==========================================" << endl << endl;

    // W1
    check(ref_build_adj_list(4, {{0,1},{0,2},{1,3}})[0] == vector<int>({1,2}), "W1: node 0");
    check(ref_build_adj_list(4, {{0,1},{0,2},{1,3}})[2] == vector<int>({0}), "W1: node 2");
    check(ref_build_adj_list(3, {})[0] == vector<int>({}), "W1: no edges");
    check(ref_build_adj_list(1, {})[0] == vector<int>({}), "W1: single node");

    // W2
    check(ref_bfs_traversal(5, {{0,1},{0,2},{1,3},{2,3},{3,4}}, 0) == vector<int>({0,1,2,3,4}), "W2: basic");
    check(ref_bfs_traversal(3, {{0,1},{1,2}}, 2) == vector<int>({2,1,0}), "W2: from end");
    check(ref_bfs_traversal(1, {}, 0) == vector<int>({0}), "W2: single");
    check(ref_bfs_traversal(4, {{0,1},{2,3}}, 0) == vector<int>({0,1}), "W2: disconnected");

    // W3
    check(ref_dfs_traversal(5, {{0,1},{0,2},{1,3},{2,3},{3,4}}, 0) == vector<int>({0,1,3,2,4}), "W3: basic");
    check(ref_dfs_traversal(3, {{0,1},{1,2}}, 0) == vector<int>({0,1,2}), "W3: linear");
    check(ref_dfs_traversal(1, {}, 0) == vector<int>({0}), "W3: single");
    check(ref_dfs_traversal(4, {{0,1},{1,2},{2,3}}, 1) == vector<int>({1,0,2,3}), "W3: from middle");

    // W4
    check(ref_count_components(5, {{0,1},{1,2},{3,4}}) == 2, "W4: two");
    check(ref_count_components(4, {}) == 4, "W4: no edges");
    check(ref_count_components(3, {{0,1},{1,2},{0,2}}) == 1, "W4: full");
    check(ref_count_components(1, {}) == 1, "W4: single");
    check(ref_count_components(7, {{0,1},{0,2},{3,4},{3,5}}) == 3, "W4: three");

    // W5
    check(ref_is_path_exists(5, {{0,1},{1,2},{3,4}}, 0, 2) == true, "W5: exists");
    check(ref_is_path_exists(5, {{0,1},{1,2},{3,4}}, 0, 4) == false, "W5: no path");
    check(ref_is_path_exists(3, {}, 0, 0) == true, "W5: same");
    check(ref_is_path_exists(3, {{0,1},{1,2}}, 0, 1) == true, "W5: direct");
    check(ref_is_path_exists(3, {{0,1}}, 0, 2) == false, "W5: isolated");

    // P1
    check(ref_shortest_path(5, {{0,1},{0,2},{1,3},{2,3},{3,4}}, 0) == vector<int>({0,1,1,2,3}), "P1: basic");
    check(ref_shortest_path(4, {{0,1},{2,3}}, 0) == vector<int>({0,1,-1,-1}), "P1: disconnected");
    check(ref_shortest_path(1, {}, 0) == vector<int>({0}), "P1: single");
    check(ref_shortest_path(4, {{0,1},{1,2},{2,3}}, 0) == vector<int>({0,1,2,3}), "P1: linear");
    check(ref_shortest_path(5, {{0,1},{1,2},{2,3},{3,4}}, 2) == vector<int>({2,1,0,1,2}), "P1: middle");

    // P2
    check(ref_detect_cycle(4, {{0,1},{1,2},{2,3}}) == false, "P2: no cycle");
    check(ref_detect_cycle(4, {{0,1},{1,2},{2,3},{3,0}}) == true, "P2: has cycle");
    check(ref_detect_cycle(3, {{0,1},{1,2},{0,2}}) == true, "P2: triangle");
    check(ref_detect_cycle(3, {}) == false, "P2: no edges");
    check(ref_detect_cycle(5, {{0,1},{2,3},{3,4},{4,2}}) == true, "P2: disc w/ cycle");
    check(ref_detect_cycle(5, {{0,1},{2,3}}) == false, "P2: disc no cycle");

    // P3
    check(ref_bipartite(4, {{0,1},{1,2},{2,3},{3,0}}) == true, "P3: even cycle");
    check(ref_bipartite(3, {{0,1},{1,2},{0,2}}) == false, "P3: triangle");
    check(ref_bipartite(3, {}) == true, "P3: no edges");
    check(ref_bipartite(2, {{0,1}}) == true, "P3: single edge");
    check(ref_bipartite(5, {{0,1},{2,3}}) == true, "P3: disc bipartite");
    check(ref_bipartite(5, {{0,1},{1,2},{2,3},{3,4},{4,0}}) == false, "P3: 5-cycle");

    // P4
    vector<vector<int>> adj4 = {{1,2},{0,3},{0,3},{1,2}};
    auto clone4 = ref_clone_graph(adj4);
    check(clone4 == adj4, "P4: content match");
    check(ref_clone_graph({}).empty(), "P4: empty");
    check(ref_clone_graph({{}}) == vector<vector<int>>({{}}), "P4: single node");

    // P5
    check(ref_all_paths(4, {{0,1},{0,2},{1,3},{2,3}}) ==
        vector<vector<int>>({{0,1,3},{0,2,3}}), "P5: basic");
    check(ref_all_paths(4, {{0,1},{0,2},{1,2},{1,3},{2,3}}) ==
        vector<vector<int>>({{0,1,2,3},{0,1,3},{0,2,3}}), "P5: multiple");
    check(ref_all_paths(2, {{0,1}}) == vector<vector<int>>({{0,1}}), "P5: direct");
    check(ref_all_paths(3, {{0,1}}).empty(), "P5: no path");

    // C1
    check(ref_num_provinces({{1,1,0},{1,1,0},{0,0,1}}) == 2, "C1: two");
    check(ref_num_provinces({{1,0,0},{0,1,0},{0,0,1}}) == 3, "C1: three");
    check(ref_num_provinces({{1,1,1},{1,1,1},{1,1,1}}) == 1, "C1: one");
    check(ref_num_provinces({{1}}) == 1, "C1: single");

    // C2
    check(ref_course_schedule(2, {{1,0}}) == true, "C2: no cycle");
    check(ref_course_schedule(2, {{1,0},{0,1}}) == false, "C2: cycle");
    check(ref_course_schedule(4, {{1,0},{2,1},{3,2}}) == true, "C2: chain");
    check(ref_course_schedule(3, {}) == true, "C2: no prereqs");
    check(ref_course_schedule(4, {{1,0},{2,1},{0,2}}) == false, "C2: complex");
    check(ref_course_schedule(4, {{1,0},{3,2}}) == true, "C2: disconnected");

    // C3
    check(ref_word_ladder("hit", "cog", {"hot","dot","dog","lot","log","cog"}) == 5, "C3: basic");
    check(ref_word_ladder("hit", "cog", {"hot","dot","dog","lot","log"}) == 0, "C3: no path");
    check(ref_word_ladder("hot", "dot", {"dot"}) == 2, "C3: direct");
    check(ref_word_ladder("a", "c", {"a","b","c"}) == 2, "C3: single letter");
    check(ref_word_ladder("abc", "xyz", {"abd","acd"}) == 0, "C3: end not in list");

    cout << endl;
    if (failed == 0) {
        cout << "All " << passed << " tests passed!" << endl;
    } else {
        cout << passed << " passed, " << failed << " failed." << endl;
        return 1;
    }
    return 0;
}

File: code/java/ch19/tests/TestCh19.java
package ch19.tests;

import java.util.*;
import ch19.solutions.*;

/**
 * Tests for Chapter 19: Graphs I — Exploring Networks
 *
 * Build and run:
 *   cd code/java
 *   javac ch19/tests/TestCh19.java
 *   java -ea ch19.tests.TestCh19
 */
public class TestCh19 {

    // ── Helper methods ──────────────────────────────────────────────

    static int passed = 0;
    static int failed = 0;

    static void assertEquals(int expected, int actual, String msg) {
        if (expected == actual) {
            passed++;
        } else {
            failed++;
            System.out.println("FAIL: " + msg + " — expected " + expected + ", got " + actual);
        }
    }

    static void assertBoolEquals(boolean expected, boolean actual, String msg) {
        if (expected == actual) {
            passed++;
        } else {
            failed++;
            System.out.println("FAIL: " + msg + " — expected " + expected + ", got " + actual);
        }
    }

    static void assertArrayEquals(int[] expected, int[] actual, String msg) {
        if (Arrays.equals(expected, actual)) {
            passed++;
        } else {
            failed++;
            System.out.println("FAIL: " + msg + " — expected " + Arrays.toString(expected) + ", got " + Arrays.toString(actual));
        }
    }

    static void assertListEquals(List<Integer> expected, List<Integer> actual, String msg) {
        if (expected.equals(actual)) {
            passed++;
        } else {
            failed++;
            System.out.println("FAIL: " + msg + " — expected " + expected + ", got " + actual);
        }
    }

    static void assertNestedListEquals(List<List<Integer>> expected, List<List<Integer>> actual, String msg) {
        if (expected.equals(actual)) {
            passed++;
        } else {
            failed++;
            System.out.println("FAIL: " + msg + " — expected " + expected + ", got " + actual);
        }
    }

    // ── W1: Build Adjacency List ────────────────────────────────────

    static void testW1() {
        List<List<Integer>> adj = Warmup01Sol.solve(4, new int[][]{{0,1},{0,2},{1,3}});
        assertListEquals(Arrays.asList(1,2), adj.get(0), "W1: node 0");
        assertListEquals(Arrays.asList(0,3), adj.get(1), "W1: node 1");
        assertListEquals(Arrays.asList(0), adj.get(2), "W1: node 2");
        assertListEquals(Arrays.asList(1), adj.get(3), "W1: node 3");

        List<List<Integer>> adj2 = Warmup01Sol.solve(3, new int[][]{});
        assertListEquals(new ArrayList<>(), adj2.get(0), "W1: no edges");

        List<List<Integer>> adj3 = Warmup01Sol.solve(1, new int[][]{});
        assertListEquals(new ArrayList<>(), adj3.get(0), "W1: single node");
    }

    // ── W2: BFS Traversal ──────────────────────────────────────────

    static void testW2() {
        assertListEquals(Arrays.asList(0,1,2,3,4),
            Warmup02Sol.solve(5, new int[][]{{0,1},{0,2},{1,3},{2,3},{3,4}}, 0), "W2: basic");
        assertListEquals(Arrays.asList(2,1,0),
            Warmup02Sol.solve(3, new int[][]{{0,1},{1,2}}, 2), "W2: from end");
        assertListEquals(Arrays.asList(0),
            Warmup02Sol.solve(1, new int[][]{}, 0), "W2: single node");
        assertListEquals(Arrays.asList(0,1),
            Warmup02Sol.solve(4, new int[][]{{0,1},{2,3}}, 0), "W2: disconnected");
    }

    // ── W3: DFS Traversal ──────────────────────────────────────────

    static void testW3() {
        assertListEquals(Arrays.asList(0,1,3,2,4),
            Warmup03Sol.solve(5, new int[][]{{0,1},{0,2},{1,3},{2,3},{3,4}}, 0), "W3: basic");
        assertListEquals(Arrays.asList(0,1,2),
            Warmup03Sol.solve(3, new int[][]{{0,1},{1,2}}, 0), "W3: linear");
        assertListEquals(Arrays.asList(0),
            Warmup03Sol.solve(1, new int[][]{}, 0), "W3: single node");
        assertListEquals(Arrays.asList(1,0,2,3),
            Warmup03Sol.solve(4, new int[][]{{0,1},{1,2},{2,3}}, 1), "W3: from middle");
    }

    // ── W4: Count Connected Components ─────────────────────────────

    static void testW4() {
        assertEquals(2, Warmup04Sol.solve(5, new int[][]{{0,1},{1,2},{3,4}}), "W4: two components");
        assertEquals(4, Warmup04Sol.solve(4, new int[][]{}), "W4: no edges");
        assertEquals(1, Warmup04Sol.solve(3, new int[][]{{0,1},{1,2},{0,2}}), "W4: fully connected");
        assertEquals(1, Warmup04Sol.solve(1, new int[][]{}), "W4: single node");
        assertEquals(3, Warmup04Sol.solve(7, new int[][]{{0,1},{0,2},{3,4},{3,5}}), "W4: three components");
    }

    // ── W5: Is Path Exists ─────────────────────────────────────────

    static void testW5() {
        assertBoolEquals(true, Warmup05Sol.solve(5, new int[][]{{0,1},{1,2},{3,4}}, 0, 2), "W5: path exists");
        assertBoolEquals(false, Warmup05Sol.solve(5, new int[][]{{0,1},{1,2},{3,4}}, 0, 4), "W5: no path");
        assertBoolEquals(true, Warmup05Sol.solve(3, new int[][]{}, 0, 0), "W5: same node");
        assertBoolEquals(true, Warmup05Sol.solve(3, new int[][]{{0,1},{1,2}}, 0, 1), "W5: direct");
        assertBoolEquals(false, Warmup05Sol.solve(3, new int[][]{{0,1}}, 0, 2), "W5: isolated");
    }

    // ── P1: Shortest Path ──────────────────────────────────────────

    static void testP1() {
        assertArrayEquals(new int[]{0,1,1,2,3},
            Practice01Sol.solve(5, new int[][]{{0,1},{0,2},{1,3},{2,3},{3,4}}, 0), "P1: basic");
        assertArrayEquals(new int[]{0,1,-1,-1},
            Practice01Sol.solve(4, new int[][]{{0,1},{2,3}}, 0), "P1: disconnected");
        assertArrayEquals(new int[]{0},
            Practice01Sol.solve(1, new int[][]{}, 0), "P1: single");
        assertArrayEquals(new int[]{0,1,2,3},
            Practice01Sol.solve(4, new int[][]{{0,1},{1,2},{2,3}}, 0), "P1: linear");
        assertArrayEquals(new int[]{2,1,0,1,2},
            Practice01Sol.solve(5, new int[][]{{0,1},{1,2},{2,3},{3,4}}, 2), "P1: from middle");
    }

    // ── P2: Detect Cycle ───────────────────────────────────────────

    static void testP2() {
        assertBoolEquals(false, Practice02Sol.solve(4, new int[][]{{0,1},{1,2},{2,3}}), "P2: no cycle");
        assertBoolEquals(true, Practice02Sol.solve(4, new int[][]{{0,1},{1,2},{2,3},{3,0}}), "P2: has cycle");
        assertBoolEquals(true, Practice02Sol.solve(3, new int[][]{{0,1},{1,2},{0,2}}), "P2: triangle");
        assertBoolEquals(false, Practice02Sol.solve(3, new int[][]{}), "P2: no edges");
        assertBoolEquals(true, Practice02Sol.solve(5, new int[][]{{0,1},{2,3},{3,4},{4,2}}), "P2: disconnected w/ cycle");
        assertBoolEquals(false, Practice02Sol.solve(5, new int[][]{{0,1},{2,3}}), "P2: disconnected no cycle");
    }

    // ── P3: Bipartite Check ────────────────────────────────────────

    static void testP3() {
        assertBoolEquals(true, Practice03Sol.solve(4, new int[][]{{0,1},{1,2},{2,3},{3,0}}), "P3: even cycle");
        assertBoolEquals(false, Practice03Sol.solve(3, new int[][]{{0,1},{1,2},{0,2}}), "P3: triangle");
        assertBoolEquals(true, Practice03Sol.solve(3, new int[][]{}), "P3: no edges");
        assertBoolEquals(true, Practice03Sol.solve(2, new int[][]{{0,1}}), "P3: single edge");
        assertBoolEquals(true, Practice03Sol.solve(5, new int[][]{{0,1},{2,3}}), "P3: disconnected bipartite");
        assertBoolEquals(false, Practice03Sol.solve(5, new int[][]{{0,1},{1,2},{2,3},{3,4},{4,0}}), "P3: 5-cycle");
    }

    // ── P4: Clone Graph ────────────────────────────────────────────

    static void testP4() {
        List<List<Integer>> adj = new ArrayList<>();
        adj.add(Arrays.asList(1,2)); adj.add(Arrays.asList(0,3));
        adj.add(Arrays.asList(0,3)); adj.add(Arrays.asList(1,2));
        List<List<Integer>> clone = Practice04Sol.solve(adj);
        assertNestedListEquals(adj, clone, "P4: basic content match");
        assertBoolEquals(true, clone != adj, "P4: different top-level ref");

        List<List<Integer>> empty = new ArrayList<>();
        assertNestedListEquals(empty, Practice04Sol.solve(empty), "P4: empty");

        List<List<Integer>> single = new ArrayList<>();
        single.add(new ArrayList<>());
        List<List<Integer>> singleClone = Practice04Sol.solve(single);
        assertNestedListEquals(single, singleClone, "P4: single node");
    }

    // ── P5: All Paths ──────────────────────────────────────────────

    static void testP5() {
        List<List<Integer>> expected1 = new ArrayList<>();
        expected1.add(Arrays.asList(0,1,3)); expected1.add(Arrays.asList(0,2,3));
        assertNestedListEquals(expected1,
            Practice05Sol.solve(4, new int[][]{{0,1},{0,2},{1,3},{2,3}}), "P5: basic");

        List<List<Integer>> expected2 = new ArrayList<>();
        expected2.add(Arrays.asList(0,1,2,3)); expected2.add(Arrays.asList(0,1,3));
        expected2.add(Arrays.asList(0,2,3));
        assertNestedListEquals(expected2,
            Practice05Sol.solve(4, new int[][]{{0,1},{0,2},{1,2},{1,3},{2,3}}), "P5: multiple");

        List<List<Integer>> expected3 = new ArrayList<>();
        expected3.add(Arrays.asList(0,1));
        assertNestedListEquals(expected3,
            Practice05Sol.solve(2, new int[][]{{0,1}}), "P5: direct");

        assertNestedListEquals(new ArrayList<>(),
            Practice05Sol.solve(3, new int[][]{{0,1}}), "P5: no path");
    }

    // ── C1: Number of Provinces ────────────────────────────────────

    static void testC1() {
        assertEquals(2, Challenge01Sol.solve(new int[][]{{1,1,0},{1,1,0},{0,0,1}}), "C1: two provinces");
        assertEquals(3, Challenge01Sol.solve(new int[][]{{1,0,0},{0,1,0},{0,0,1}}), "C1: three provinces");
        assertEquals(1, Challenge01Sol.solve(new int[][]{{1,1,1},{1,1,1},{1,1,1}}), "C1: one province");
        assertEquals(1, Challenge01Sol.solve(new int[][]{{1}}), "C1: single");
    }

    // ── C2: Course Schedule ────────────────────────────────────────

    static void testC2() {
        assertBoolEquals(true, Challenge02Sol.solve(2, new int[][]{{1,0}}), "C2: no cycle");
        assertBoolEquals(false, Challenge02Sol.solve(2, new int[][]{{1,0},{0,1}}), "C2: cycle");
        assertBoolEquals(true, Challenge02Sol.solve(4, new int[][]{{1,0},{2,1},{3,2}}), "C2: chain");
        assertBoolEquals(true, Challenge02Sol.solve(3, new int[][]{}), "C2: no prereqs");
        assertBoolEquals(false, Challenge02Sol.solve(4, new int[][]{{1,0},{2,1},{0,2}}), "C2: complex cycle");
        assertBoolEquals(true, Challenge02Sol.solve(4, new int[][]{{1,0},{3,2}}), "C2: disconnected");
    }

    // ── C3: Word Ladder ────────────────────────────────────────────

    static void testC3() {
        assertEquals(5, Challenge03Sol.solve("hit", "cog",
            Arrays.asList("hot","dot","dog","lot","log","cog")), "C3: basic");
        assertEquals(0, Challenge03Sol.solve("hit", "cog",
            Arrays.asList("hot","dot","dog","lot","log")), "C3: no path");
        assertEquals(2, Challenge03Sol.solve("hot", "dot",
            Arrays.asList("dot")), "C3: direct");
        assertEquals(2, Challenge03Sol.solve("a", "c",
            Arrays.asList("a","b","c")), "C3: single letter");
        assertEquals(0, Challenge03Sol.solve("abc", "xyz",
            Arrays.asList("abd","acd")), "C3: end not in list");
    }

    // ── Main ────────────────────────────────────────────────────────

    public static void main(String[] args) {
        System.out.println("Chapter 19: Graphs I — Exploring Networks");
        System.out.println("==========================================\n");

        testW1();
        testW2();
        testW3();
        testW4();
        testW5();
        testP1();
        testP2();
        testP3();
        testP4();
        testP5();
        testC1();
        testC2();
        testC3();

        System.out.println();
        if (failed == 0) {
            System.out.println("All " + passed + " tests passed!");
        } else {
            System.out.println(passed + " passed, " + failed + " failed.");
            System.exit(1);
        }
    }
}
